File: app/src/main/java/com/sdex/activityrunner/MainActivity.java
package com.sdex.activityrunner;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Build.VERSION;
import android.os.Build.VERSION_CODES;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.ContentLoadingProgressBar;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.SearchView.OnQueryTextListener;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;

import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClient.BillingResponse;
import com.android.billingclient.api.BillingClient.SkuType;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.Purchase.PurchasesResult;
import com.codemybrainsout.ratingdialog.RatingDialog;
import com.sdex.activityrunner.intent.IntentBuilderActivity;
import com.sdex.activityrunner.preferences.SettingsActivity;
import com.sdex.activityrunner.service.AppLoaderIntentService;
import com.sdex.activityrunner.util.RecyclerViewHelper;
import com.sdex.commons.BaseActivity;
import com.sdex.commons.ads.AdsDelegate;
import com.sdex.commons.ads.AppPreferences;
import com.sdex.commons.util.AppUtils;
import com.sdex.commons.util.UIUtils;

import java.util.List;

public class MainActivity extends BaseActivity {

  private static final String STATE_SEARCH_TEXT = "state_search_text";

  private AdsDelegate adsDelegate;
  private AppPreferences appPreferences;
  private boolean isProVersionEnabled;
  private ApplicationsListAdapter adapter;
  private SwipeRefreshLayout refreshLayout;
  private ContentLoadingProgressBar progressBar;
  private ApplicationListViewModel viewModel;
  private String searchText;

  @Override
  protected int getLayout() {
    return R.layout.activity_main;
  }

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    AppLoaderIntentService.enqueueWork(this, new Intent());

    viewModel = ViewModelProviders.of(this).get(ApplicationListViewModel.class);

    appPreferences = new AppPreferences(this);
    adsDelegate = new AdsDelegate(appPreferences);
    adsDelegate.initInterstitialAd(this, R.string.ad_interstitial_unit_id);

    fetchPurchases();
    showRatingDialog();

    if (savedInstanceState != null) {
      searchText = savedInstanceState.getString(STATE_SEARCH_TEXT);
    }

    progressBar = findViewById(R.id.progress);
    progressBar.show();
    refreshLayout = findViewById(R.id.refresh);
    RecyclerView list = findViewById(R.id.list);
    RecyclerViewHelper.addDivider(list);
    adapter = new ApplicationsListAdapter(this);
    list.setAdapter(adapter);
    refreshLayout.setOnRefreshListener(() -> {
      refreshLayout.setRefreshing(true);
      final Intent work = new Intent();
      work.putExtra(AppLoaderIntentService.ARG_REASON, AppLoaderIntentService.REFRESH_USER);
      AppLoaderIntentService.enqueueWork(this, work);
    });

    checkOreoBug();
  }

  @Override
  public void onSaveInstanceState(@NonNull Bundle outState) {
    super.onSaveInstanceState(outState);
    outState.putString(STATE_SEARCH_TEXT, searchText);
  }

  @Override
  public void onResume() {
    super.onResume();
    viewModel.getItems(searchText).observe(this, itemModels -> {
      if (itemModels != null) {
        adapter.submitList(itemModels);
        refreshLayout.setRefreshing(false);
        progressBar.hide();
      }
    });
  }

  private void filter(String text) {
    if (adapter != null) {
      this.searchText = text;
      viewModel.getItems(text).observe(this,
        itemModels -> adapter.submitList(itemModels));
    }
  }

  // https://issuetracker.google.com/issues/73289329
  private void checkOreoBug() {
    if (VERSION.SDK_INT == VERSION_CODES.O) {
      boolean warningWasShown = appPreferences.getPreferences().getBoolean(
        OreoPackageManagerBugActivity.KEY, false);
      if (!warningWasShown) {
        MainViewModel viewModel = ViewModelProviders.of(this).get(MainViewModel.class);
        viewModel.getPackages().observe(this, packageInfo -> {
          if (packageInfo != null) {
            if (packageInfo.isEmpty()) {
              overridePendingTransition(0, 0);
              startActivity(new Intent(this, OreoPackageManagerBugActivity.class));
            }
          }
        });
      }
    }
  }

  private void fetchPurchases() {
    BillingClient billingClient = BillingClient.newBuilder(this)
      .setListener((responseCode, purchases) -> {
        if (responseCode == BillingResponse.OK) {
          handlePurchases(purchases);
        }
      })
      .build();
    billingClient.startConnection(new BillingClientStateListener() {
      @Override
      public void onBillingSetupFinished(@BillingResponse int billingResponseCode) {
        if (billingResponseCode == BillingResponse.OK) {
          PurchasesResult purchasesResult = billingClient.queryPurchases(SkuType.INAPP);
          List<Purchase> purchases = purchasesResult.getPurchasesList();
          handlePurchases(purchases);
        }
      }

      @Override
      public void onBillingServiceDisconnected() {
        // Try to restart the connection on the next request to
        // Google Play by calling the startConnection() method.
      }
    });
  }

  private void handlePurchases(@Nullable List<Purchase> purchases) {
    if (purchases != null) {
      for (Purchase purchase : purchases) {
        if (PurchaseActivity.isPremiumVersion(purchase.getSku())) {
          isProVersionEnabled = true;
          invalidateOptionsMenu();
          break;
        }
      }
    }
    appPreferences.setProVersion(isProVersionEnabled);
  }

  private void showRatingDialog() {
    final RatingDialog ratingDialog = new RatingDialog.Builder(this)
      .threshold(3)
      .session(15)
      .onRatingBarFormSumbit(feedback ->
        AppUtils.sendEmail(this, AppUtils.ACTIVITY_RUNNER_FEEDBACK_EMAIL,
          AppUtils.ACTIVITY_RUNNER_FEEDBACK_SUBJECT, feedback))
      .build();
    ratingDialog.show();
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.main, menu);
    MenuItem searchItem = menu.findItem(R.id.action_search);
    SearchView searchView = (SearchView) searchItem.getActionView();
    String hint = getString(R.string.action_search_hint);
    searchView.setQueryHint(hint);

    if (!TextUtils.isEmpty(searchText)) {
      searchView.post(() -> searchView.setQuery(searchText, false));
      searchItem.expandActionView();
      UIUtils.setMenuItemsVisibility(menu, searchItem, false);
    }

    searchView.setOnQueryTextListener(new OnQueryTextListener() {
      @Override
      public boolean onQueryTextSubmit(String query) {
        return false;
      }

      @Override
      public boolean onQueryTextChange(String newText) {
        filter(newText);
        return false;
      }
    });
    searchItem.setOnActionExpandListener(new MenuItem.OnActionExpandListener() {
      @Override
      public boolean onMenuItemActionExpand(MenuItem item) {
        UIUtils.setMenuItemsVisibility(menu, item, false);
        return true;
      }

      @Override
      public boolean onMenuItemActionCollapse(MenuItem item) {
        UIUtils.setMenuItemsVisibility(menu, true);
        invalidateOptionsMenu();
        return true;
      }
    });
    return super.onCreateOptionsMenu(menu);
  }

  @Override
  public boolean onPrepareOptionsMenu(Menu menu) {
    if (isProVersionEnabled) {
      menu.findItem(R.id.action_upgrade).setVisible(false);
    }
    return super.onPrepareOptionsMenu(menu);
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.action_launch_intent: {
        IntentBuilderActivity.start(this, null);
        return true;
      }
      case R.id.action_upgrade: {
        PurchaseActivity.start(this);
        return true;
      }
      case R.id.action_about: {
        AboutActivity.start(this);
        return true;
      }
      case R.id.action_settings: {
        SettingsActivity.start(this, SettingsActivity.NORMAL);
        return true;
      }
      default:
        return super.onOptionsItemSelected(item);
    }
  }

  @Override
  public void onBackPressed() {
    super.onBackPressed();
    adsDelegate.showInterstitial();
  }
}

File: app/src/main/java/com/sdex/activityrunner/MainActivity.kt
package com.sdex.activityrunner

import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatDelegate
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.SearchView.OnQueryTextListener
import androidx.core.view.MenuProvider
import androidx.core.view.isVisible
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.sdex.activityrunner.about.AboutActivity
import com.sdex.activityrunner.app.ActivitiesListActivity
import com.sdex.activityrunner.app.ApplicationsListAdapter
import com.sdex.activityrunner.app.MainViewModel
import com.sdex.activityrunner.app.dialog.ApplicationOptionsDialog
import com.sdex.activityrunner.commons.BaseActivity
import com.sdex.activityrunner.databinding.ActivityMainBinding
import com.sdex.activityrunner.db.cache.ApplicationModel
import com.sdex.activityrunner.extensions.setItemsVisibility
import com.sdex.activityrunner.intent.IntentBuilderActivity
import com.sdex.activityrunner.preferences.AppPreferences
import com.sdex.activityrunner.preferences.DisplayConfig
import com.sdex.activityrunner.preferences.PreferencesBottomDialog
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class MainActivity : BaseActivity() {

    @Inject
    lateinit var appPreferences: AppPreferences
    private val viewModel by viewModels<MainViewModel>()

    private lateinit var binding: ActivityMainBinding
    private lateinit var adapter: ApplicationsListAdapter

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setupToolbar()

        addMenuProvider(
            object : MenuProvider {
                override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                    menuInflater.inflate(R.menu.main, menu)
                    configureSearchView(menu)
                }

                override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                    return when (menuItem.itemId) {
                        R.id.action_settings -> {
                            PreferencesBottomDialog().show(
                                supportFragmentManager,
                                PreferencesBottomDialog.TAG,
                            )
                            true
                        }

                        R.id.action_launch_intent -> {
                            IntentBuilderActivity.start(this@MainActivity, null)
                            true
                        }

                        R.id.action_about -> {
                            AboutActivity.start(this@MainActivity)
                            true
                        }

                        else -> false
                    }
                }
            },
        )

        adapter = ApplicationsListAdapter(this, DisplayConfig()).apply {
            itemClickListener = object : ApplicationsListAdapter.ItemClickListener {
                override fun onItemClick(item: ApplicationModel) {
                    ActivitiesListActivity.start(this@MainActivity, item)
                }

                override fun onItemLongClick(item: ApplicationModel) {
                    val dialog = ApplicationOptionsDialog.newInstance(item)
                    dialog.show(supportFragmentManager, ApplicationOptionsDialog.TAG)
                }
            }
        }

        binding.list.adapter = adapter
        binding.list.setHasFixedSize(true)

        lifecycleScope.launch {
            viewModel.uiState.flowWithLifecycle(lifecycle)
                .collect { state ->
                    adapter.updateDisplayConfig(state.displayConfig)
                    adapter.submitList(state.items) {
                        if (state.items.isNotEmpty()) {
                            binding.progress.hide()
                        }
                        if (shouldScrollToTop()) {
                            binding.list.scrollToPosition(0)
                        }
                    }
                }
        }

        lifecycleScope.launch {
            viewModel.isSyncing.flowWithLifecycle(lifecycle)
                .collect { isSyncing ->
                    binding.syncProgress.isVisible = isSyncing
                }
        }

        binding.progress.show()
    }

    override fun onResume() {
        super.onResume()

        viewModel.quickSync()
    }

    private fun shouldScrollToTop(): Boolean {
        // scroll to top when the filter dialog is shown
        return supportFragmentManager.findFragmentByTag(PreferencesBottomDialog.TAG) != null ||
            // scroll to top when the list already is at the top to display new items
            (binding.list.layoutManager as LinearLayoutManager).findFirstCompletelyVisibleItemPosition() == 0
    }

    private fun configureSearchView(menu: Menu) {
        val searchItem = menu.findItem(R.id.action_search)
        val searchView = searchItem.actionView as SearchView
        // expand the view to the full width: https://stackoverflow.com/a/34050959/2894324
        searchView.apply {
            maxWidth = Int.MAX_VALUE
            queryHint = getString(R.string.action_search_hint)
        }

        val searchQuery = viewModel.searchQuery.value
        if (!searchQuery.isNullOrEmpty()) {
            searchView.post { searchView.setQuery(searchQuery, false) }
            searchItem.expandActionView()
            menu.setItemsVisibility(exception = searchItem, visible = false)
        }

        searchView.setOnQueryTextListener(
            object : OnQueryTextListener {
                override fun onQueryTextSubmit(query: String): Boolean {
                    return false
                }

                override fun onQueryTextChange(newText: String): Boolean {
                    viewModel.search(newText)
                    return false
                }
            },
        )
        searchItem.setOnActionExpandListener(
            object : MenuItem.OnActionExpandListener {
                override fun onMenuItemActionExpand(item: MenuItem): Boolean {
                    menu.setItemsVisibility(exception = item, visible = false)
                    return true
                }

                override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
                    menu.setItemsVisibility(visible = true)
                    invalidateOptionsMenu()
                    return true
                }
            },
        )
    }

    companion object {

        init {
            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        }
    }
}
